fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/CalculateIncomeAndExpenseBookingIntegrationTest.java

```diff
@@ -0,0 +1,161 @@
+
+    private JsonCommand jsonCommandWrapperTest;
+    private IncomeAndExpenseReadPlatformService incomeAndExpenseReadPlatformService;
+    private OfficeReadPlatformService officeReadPlatformService;
+    private CalculateIncomeAndExpenseBookingImpl calculateIncomeAndExpenseBooking;
+
+    @Before
+    public void setup() {
+        calculateIncomeAndExpenseBooking =  new CalculateIncomeAndExpenseBookingImpl(null, null, null, null, incomeAndExpenseReadPlatformService,officeReadPlatformService);
+    }
+
+    public JsonCommand getCommand() {
+        return jsonCommandWrapperTest;
+    }
+
+    public JsonCommand setCommand(JsonCommand command) {
+        return this.jsonCommandWrapperTest = command;
+    }
+
+    /*
+    Case 1: All running balances has to be calculated before booking off income and expense account
+    If not running balances, then throw exception
+     */
+    @Test(expected = RunningBalanceNotCalculatedException.class)
+    public void testBookOffIncomeAndExpenseCheckException() {
+        GLClosureCommand glClosureCommand =  new GLClosureCommand((long)10,(long)10, new LocalDate(),"Closing comment",  false, (long)10 , "CAD", false, false, "comment" );
+        IncomeAndExpenseJournalEntryData incomeAndExpenseJournalEntryData= new IncomeAndExpenseJournalEntryData(null, null,null, null
+                , true, false, null,null,null,10,10,null,null);
+        List<IncomeAndExpenseJournalEntryData> incomeAndExpenseJournalEntryDataList = new ArrayList<>();
+        incomeAndExpenseJournalEntryDataList.add(incomeAndExpenseJournalEntryData);
+        calculateIncomeAndExpenseBooking.bookOffIncomeAndExpense(incomeAndExpenseJournalEntryDataList,glClosureCommand, false,null,null);
+    }
+
+    /*
+    Case 2: All running balances has to be calculated before booking off income and expense account
+    No exception is to be thrown since it is running for the input.
+    However, later a null pointer exception is thrown, since the object incomeAndExpense is not initialized.
+     */
+    @Test
+    public void testBookOffIncomeAndExpense() {
+        GLClosureCommand glClosureCommand =  new GLClosureCommand(10L, 10L, new LocalDate(), "Closing comment", false, 10L , "CAD", false, false, "comment" );
+        IncomeAndExpenseJournalEntryData incomeAndExpenseJournalEntryData= new IncomeAndExpenseJournalEntryData(null, null,null, null
+                , true, true, null,null,null,10,10,null,null);
+        List<IncomeAndExpenseJournalEntryData> incomeAndExpenseJournalEntryDataList = new ArrayList<>();
+        incomeAndExpenseJournalEntryDataList.add(incomeAndExpenseJournalEntryData);
+        Assert.assertNotNull(calculateIncomeAndExpenseBooking.bookOffIncomeAndExpense(incomeAndExpenseJournalEntryDataList, glClosureCommand, false,null,null));
+    }
+
+    /*
+    Case 3: In the case of an income account type-  if the OfficeRunningBalance is greater than 0, then add to debits
+     */
+    @Test(expected = NoClassDefFoundError.class)
```

Comment:
This (`@Test(expected = NoClassDefFoundError.class)`) seems really weird to me, at first review. Not the expecting an exception, that is fine, as above in `testBookOffIncomeAndExpenseCheckException` for `RunningBalanceNotCalculatedException`, but why would a "functional" test ever hit a `NoClassDefFoundError`, which is a (very) low-level technical exception that (normally) "should never happen"?